Code for analyse step with end-to-end test

pchlap opened the pull request:
Hi @rnfinnegan and @dalmouiee,

I've been working with pydicer to extract some data and compute GTV volumes of the structures for Farhannah... And it made me realize this kind of analysis on the converted data is going to be super common. I was computing radiomics which can compute the volume but also many other features. So I decided to add it in to pydicer as a potential final step. We can also add DVH analysis to this and it will make all of these tasks super easy in the future.

I have also finally put in an end-to-end test. It could probably do some more checks once it finishes, but for now it actually just checks the values of some radiomics features which were computed. If something changes anywhere in the pipeline, it is likely that these feature values will change and will let us know. We won't yet know why or where the change actually happened so we'd need to investigate, but perhaps we can improve the end-to-end test later so that it can give us a better indication.

I'd be very curious to hear what you think about this contribution. Cheers.

dalmouiee commented on pydicer/analyse/data.py:
```diff
@@ -0,0 +1,261 @@
+            dfs.append(pd.read_csv(radiomics_file, index_col=0))
+
+        df = pd.concat(dfs)
+        df.sort_values(["Patient", "Contour"], inplace=True)
```
Is this sort order conventional and could user input on potential ordering be useful?

pchlap replied:
I'd imagine that usually you would want the output grouped by patient first and then by the contour... And this function returns a dataframe so the user can just sort that themselves if they want something different?

dalmouiee commented on pydicer/analyse/data.py:
```diff
@@ -0,0 +1,261 @@
+                )
+            patient = [patient]
+
+        if radiomics is None:
```
Could we set these default values to be in the function definition, rather than having ```None``` as the default then checking if these are ```None```? So like:
```python
def compute_radiomics(
        self,
        patient=None,
        radiomics=DEFAULT_RADIOMICS,
        settings=PYRAD_DEFAULT_SETTINGS,
        structure_meta_data=[],
        image_meta_data=[],
        resample_to_image=False,
    ):
```

pchlap replied:
I did have that, but pylint gives a warning that that isn't secure for some reason... Not actually sure why but I thought I'd just make pylint happy for now :D Will look into what exactly the reason is :)

pchlap replied:
Ah ok I see why its bad practice. So if you use a dict or list (both mutable) as a default parameter, then this same default object uses the same object across multiple calls to the function. So if it changes during the function call then it will have the changed object in the next function all instead of starting with a fresh default object.

This example should demonstrate it:
```python
def func_bad(k, v, para={}):
    para[k] = v
    return para
    
print(func_bad('a', 1)) #returns {'a': 1}
print(func_bad('b', 2)) #returns {'a': 1, 'b': 2}
print(func_bad('c', 3)) #returns {'a': 1, 'b': 2, 'c': 3}
    
    
def func_good(k, v, para=None):
    if para is None:
        para = dict()
    para[k] = v
    return para
    
print(func_good('a', 1)) #returns {'a': 1}
print(func_good('b', 2)) #returns {'b': 2}
print(func_good('c', 3)) #returns {'c': 3}
```

dalmouiee replied:
Ah interesting never knew that, great to know it now :D

dalmouiee commented on pydicer/analyse/data.py:
```diff
@@ -0,0 +1,261 @@
+                    if interpolator is not None and resample_pixel_spacing is not None:
+                        image, mask = imageoperations.resampleImage(image, mask, **settings)
+
+                    # output[contour_name] = {"Contour": contour_name}
```
Remove comment?

dalmouiee reviewed (approved):
Hey @pchlap, great stuff would be great if we can expand on the testing like you mentioned. I add some general comments for potential styling. We'll definitely discuss this further in the next catchup. Otherwise, LGTM! :)

dalmouiee commented:
Seems that the checks aren't happy with ```pip install pyradiomics```, maybe move ```numpy``` ahead of it in the reqs.txt list?

rnfinnegan reviewed (approved):
This looks good! A minor comment: it's pretty usual to do some pre-processing to get radiomics features (e.g. normalisation and discretisation). Maybe we could add these as options in a future version.

pchlap commented:
> This looks good! A minor comment: it's pretty usual to do some pre-processing to get radiomics features (e.g. normalisation and discretisation). Maybe we could add these as options in a future version.

Thanks Rob, I'm using pyradiomics under the hood so I would imagine that takes care of those sort of things?

pchlap commented:
> Seems that the checks aren't happy with `pip install pyradiomics`, maybe move `numpy` ahead of it in the reqs.txt list?

Yeah it's weird. We don't explicitly list a numpy version because it's installed via other libraries. I'll try it out though :)

dalmouiee reviewed (approved):
Great job @pchlap, the commit messages and the mentioned issues were for helping me understand the new changes thanks for that! I'm happy to merge this through for the addressed issues in the PR and will proceed with the poetry integration once this has been merged through :)

pchlap commented:
> Great job @pchlap, the commit messages and the mentioned issues were great to help me understand the new changes thanks for that! I'm happy to merge this through for the addressed issues in the PR and will proceed with the poetry integration once this has been merged through :)

Thanks @dalmouiee. Sounds good. I'll merge this through and then we'll go from there :)
